### ISSUE TITLE
[Experimental] API for layout-stability of variation axes changes?

### ISSUE THREAD

**Issue by drott** [harfbuzz/harfbuzz#4639]
From an explorative chat between @behdad and @drott:

If we were able to tell whether a variation axis change affects layout or not (important metrics: horizontal or vertical advance width, ascent, descent & glyph variations: rvrn...other?) then there is the potential to skip layout and animate only the glyph shape, allowing a rendering engine to skip layout for a variation axis change. 

@behdad mentions it's theoretically possible to add API to HarfBuzz to determine whether axes trigger layout-sensitive changes. 

Leaving this here to track, if we decide to come back to it.

**Comment by behdad** [comment]
I agree it probably be very hard for platforms to embrace such API. Leaving it around for a while to see if I can get to it.

**Comment by behdad** [comment]
What would an API for this look like possibly?

**Comment by drott** [comment]
I think the main utility would be knowing about layout stability independent of `hb_buffer_t` content, otherwise this probably gets too slow or too fine-grained to be useful. So then I believe the input would be just two sets of variation axis parameters and the response would give information about whether those two coordinate groups would affect and vary/change any advance widths and GPOS positioning deltas? Is that feasible? Return value could be an enum describing no-change, no-advance-width-change, no-advance-and-pos-change at all... or something like that?

**Comment by behdad** [comment]
I'm more open to just returning `no layout change` / `possible layout change` per variation axis. That's more feasible. And at a face level. That would look into all of GSUB/GPOS/HVAR for variations involving that axis. Would you like to try this?

@jfkthame Is this something you are interested in?

**Comment by drott** [comment]
At the moment I am afraid we don't have the machinery in the style and layout engine to make efficient use of this. Changing `font-variation-settings` values would immediately trigger a style resolution and layout. I think this feature would have great potential for speeding up variation animations, but I unfortunately do not have the cycles at the moment to prototype that. 

> I'm more open to just returning no layout change / possible layout change per variation axis. That's more feasible. And at a face level. That would look into all of GSUB/GPOS/HVAR for variations involving that axis. 

This sounds reasonable. Yes, this smaller granularity I guess could be used to compute the comparison of two variation coordinate groups for each axis and find out if all are `no-layout-change`. 

Would this `no-layout-change` also cover side-effects of one axis triggering others through `AVAR` 2?

**Comment by jfkthame** [comment]
Yeah, I'm not convinced we'd get much value out of this. In principle I can imagine using such an API to optimize some style-change handling, but I wonder how often it'd help in practice. My gut feeling is that the overwhelming majority of variation changes will result in some kind of layout change -- after all, the most commonly used axes are probably weight, width, and optical size, all of which will almost always affect layout -- so I'm concerned that we'd be doing this (somewhat complex) evaluation for all variable fonts, yet only rarely getting any benefit.

(Of course, if someone has data showing that it's actually quite common to change variation-font settings in ways that *don't* affect layout, so that the potential optimization here would be a real win, that'd make such an API more compelling.)
